### PR TITLE
1208 dashboard loading and error state

### PR DESCRIPTION
**Pull request by tcaiger**
### Changes:

- Add Alert component
- Add loading state to date picker
- Add info and error alert to Enlarged Dialog component
- Add info and error alert to dashboards

---

### Screenshots:
Screenshots here: https://github.com/beyondessential/tupaia-backlog/issues/1208

### CODE REVIEW

File: packages/web-frontend/config/webpack.config.dev.js
Comment (tcaiger) [inline packages/web-frontend/config/webpack.config.dev.js:90]: This prevents the package being loaded more than once

File: packages/web-frontend/src/actions.js
Comment (tcaiger) [inline packages/web-frontend/src/actions.js:689]: Fix existing bug. The error that gets passed from the saga is a string

File: packages/web-frontend/src/sagas.js
Comment (tcaiger) [inline packages/web-frontend/src/sagas.js:618]: Always use the designated error handler for the saga. Otherwise another error handler from the saga could handle the error.

File: packages/web-frontend/src/components/Alert.js
Comment (IgorNadj) [inline packages/web-frontend/src/components/Alert.js:89]: How come this is configurable? Seems ok to leave it up to the component to decide this.

File: packages/web-frontend/src/components/Alert.js
Comment (IgorNadj) [inline packages/web-frontend/src/components/Alert.js:75]: Storybook story for this would be great

File: packages/web-frontend/config/webpack.config.dev.js
Comment (IgorNadj) [reply packages/web-frontend/config/webpack.config.dev.js:90]: Nice, add as a comment

File: packages/web-frontend/src/sagas.js
Comment (IgorNadj) [reply packages/web-frontend/src/sagas.js:618]: I'm not a fan of ignoring errorFunctions here. errorFunctions should be checked first and handled globally. E.g. if the user is not logged in, that should be a single UI element saying you're logged out, please log in again, instead of each component on the screen having an error message saying "please log in". In that case, a simple "something went wrong" message on each component, with maybe a retry button would be fine.

It's only if the global error function is not hit that we can assume this is a user error, e.g. wrong dates, and the container that made the request can display the specific error.

```
if (error.errorFunction) {	   
  // global error (e.g. not logged in), handle globally, plus display simple error message in each component
  yield put(error.errorFunction(error, infoViewKey));	
  if (errorHandler) {
    errorHandler('Failed to load data');
  }
} else {
  // user error e.g. bad dates
  if (errorHandler) {
    const json = ...
    ...
    errorHandler(errorMessage, infoViewKey);
  }
}

```

With 500 errors:
Sometimes the server falls over and gives a 500. Usually users cannot do anything about this e.g. server is offline, or there is a bug in the code, apart from try again. 

Generally speaking, an API that returns 500 should not return any information about the error. The server fell over, and it shouldn't have. Instead it should be logged and devs alerted and go fix the error. 

It seems from the screenshot before ("Internal server error: Start period must be equal to or earlier than the end period") that some of our APIs are returning 500 with user errors, this is a problem... these should be 400 Bad Request. If that's the case then that would be a separate ticket to change these.

**Review by IgorNadj — CHANGES_REQUESTED** [review]

File: packages/web-frontend/src/components/Alert.js
Comment (tcaiger) [reply packages/web-frontend/src/components/Alert.js:89]: This is to set our own defaults over the top of the material-ui ones to match the designs. The component will also be able to customise it when it's being used if needed.

File: packages/web-frontend/src/components/Alert.js
Comment (tcaiger) [reply packages/web-frontend/src/components/Alert.js:75]: Yes good call. I have added it.

**Comment by IgorNadj** [comment]
LGTM! Nice one

**Review by IgorNadj — APPROVED** [review]

File: packages/web-frontend/src/sagas.js
Comment (IgorNadj) [inline packages/web-frontend/src/sagas.js:619]: Ah this should probably stay as errorFunction(error, infoViewKey) no?

And you removed errorHandler()? I have approved anyway, assuming it works as you want it to

Comment (tcaiger) [reply packages/web-frontend/src/sagas.js:619]: yep it is working as we discussed. On line 609 the errorHandler gets passed into the request call and then the request handler returns the correct errorHandler back in the error response. If there is a server error, then a generic errorHandler gets passed back otherwise the specified error handler gets passed. I did a bit of digging and testing and realised that this behaviour is already set up in the saga request handler.

Comment (tcaiger) [reply packages/web-frontend/src/sagas.js:619]: > Ah this should probably stay as errorFunction(error, infoViewKey) no?
> 
> And you removed errorHandler()? I have approved anyway, assuming it works as you want it to

No here I dod want errorMessage as this will give me the full error message from the server if there is a user error message present. Thanks for the review, I'm much happier with the code and my understanding than I was before :)

**Pull request merged**
